### PR TITLE
Fix zero_grad place resulting in zero logs

### LINKED ISSUE #2459: GradsScalarHandler logs 0 gradients if default update function is used
## 🐛 Bug description

Logging the gradients per epoch / iteration is a useful way to debug an under-performing model. Ignite provides an easy-to-use tensorboard_logger handler, an example accessible from `ignite.contrib.handlers.tensorboard_logger.GradsScalarHandler`. However, the default `update` function used by Engines generated by `create_supervised_trainer` zero the gradients before terminating, causing the handler to log zeroed out gradients all the time. 

**Steps to reproduce:**
My code is too complicated at the momemt to provide a clear insight, and I am limited by time to provide a minimal (not-)working example, so I will provide abstracted steps.

1. Generate an `Engine` / `DeterministicEngine` on an arbitrary problem by the `create_supervised_trainer` method.
2. Establish a `TensorboardLogger` and a .
3. Attach a `GradsScalarHandler` / your choice of a gradient logger. Also log the training loss or some other metric.
4. Start the training, check tensorboard and see the constant-0 gradient norms / gradients, despite the losses/metrics implying some sort of improvement/learning takes place.

**Solution proposal**

Preserving the gradients until epoch end is tricky, but not required for my purposes. If we are OK with using  `Events.ITERATION_COMPLETED` as a cue to log gradients, then we can simply modify the default `update` functions as follows:

(assuming  `engine.state.iteration` counts from 1).
```
    def update(engine: Engine, batch: Sequence[torch.Tensor]) -> Union[Any, Tuple[torch.Tensor]]:
        if (engine.state.iteration + 1) % gradient_accumulation_steps == 0:
            optimizer.zero_grad()
        model.train()
        x, y = prepare_batch(batch, device=device, non_blocking=non_blocking)
        y_pred = model(x)
        loss = loss_fn(y_pred, y)
        if gradient_accumulation_steps > 1:
            loss = loss / gradient_accumulation_steps
        loss.backward()
        if engine.state.iteration % gradient_accumulation_steps == 0:
            optimizer.step()
        return output_transform(x, y, y_pred, loss)
```

This way, upon completion of `update` and at the moment of `Events.ITERATION_COMPLETED` firing, there will be some non-zero gradients available to be logged.

## Environment (latest version of Ignite still has the same bug)

- PyTorch Version (e.g., 1.4): 1.10.1
- Ignite Version (e.g., 0.3.0): 0.4.7
- OS (e.g., Linux):
- How you installed Ignite (`conda`, `pip`, source): conda
- Python version: 3.9.7
- Any other relevant information:

### PR DESCRIPTION
**Pull request by sadra-barikbin**
Fixes #2459 

Description:
Based on PR #2470 . Tests are added.

Check list:

- [x] New tests are added (if a new feature is added)
- [ ] New doc strings: description and/or example code are in RST format
- [ ] Documentation is updated (if required)

### CODE REVIEW

File: tests/ignite/engine/test_create_supervised.py [hunk 1]
```diff
@@ -124,6 +124,37 @@ def _():
...
+        data = [(_x, _y) for _x, _y in zip(x, y)]
+
+        @trainer.on(Events.ITERATION_COMPLETED)
+        def _():
+            assert model.weight.data.grad != 0
+
+        trainer.run(data)
+
+    _test(1)
+    _test(3)
```
Comment (vfdev-5) [inline tests/ignite/engine/test_create_supervised.py:155]: Can you please use `@pytest.mark.parametrize` for that directly instead of calling `_test(1)`, `_test(3)`

File: tests/ignite/contrib/handlers/conftest.py [hunk 2]
```diff
@@ -1,6 +1,5 @@
 from unittest.mock import Mock
 
-import numpy as np
```
Comment (vfdev-5) [inline tests/ignite/contrib/handlers/conftest.py:3]: This should be reverted

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the PR @sadra-barikbin 
Few comments to make it better

File: tests/ignite/engine/test_create_supervised.py [hunk 3]
```diff
@@ -340,25 +369,39 @@ def _test_create_evaluation_step(
     assert output_transform_mock.called
 
 
-def test_create_supervised_trainer():
+@pytest.mark.parametrize("gradient_accumulation_steps", [1, 3])
```
Comment (vfdev-5) [inline tests/ignite/engine/test_create_supervised.py:372]: I do not understand why you changed this ?

File: tests/ignite/engine/test_create_supervised.py [hunk 4]
```diff
@@ -340,25 +369,39 @@ def _test_create_evaluation_step(
...
+def test_create_supervised_trainer(gradient_accumulation_steps):
     _test_create_supervised_trainer_wrong_accumulation()
-    _test_create_supervised_trainer(gradient_accumulation_steps=1)
-    _test_create_supervised_trainer(gradient_accumulation_steps=3)
+    _test_create_supervised_trainer(gradient_accumulation_steps=gradient_accumulation_steps)
+    _test_create_supervised_trainer(gradient_accumulation_steps=gradient_accumulation_steps)
```
Comment (vfdev-5) [inline tests/ignite/engine/test_create_supervised.py:376]: Here you call the same test twice ?

File: tests/ignite/engine/test_create_supervised.py (hunk 3 above)
Comment (sadra-barikbin) [reply tests/ignite/engine/test_create_supervised.py:372]: You mean there is no need to `_test_create_supervised_trainer_have_grad_after_iteration` with different `gradient_accumulation_steps`?

Comment (sadra-barikbin) [reply tests/ignite/engine/test_create_supervised.py:372]: @vfdev-5 You mean there is no need to _test_create_supervised_trainer_have_grad_after_iteration with different gradient_accumulation_steps?

Comment (vfdev-5) [reply tests/ignite/engine/test_create_supervised.py:372]: @sadra-barikbin try to execute these tests and see what this code do.
It will generate 2 tests like 
`test_create_supervised_trainer[gradient_accumulation_steps-1]` and `test_create_supervised_trainer[gradient_accumulation_steps-3]` where 
`_test_create_supervised_trainer_wrong_accumulation()` and `_test_create_mocked_supervised_trainer()` will be executed twice which is useless.

I suggest to revert the change and just call 
```
_test_create_supervised_trainer_have_grad_after_iteration(1)
_test_create_supervised_trainer_have_grad_after_iteration(3)
```

Another solution could be split `test_create_supervised_trainer` into 2 tests such that the first test is executing once `_test_create_supervised_trainer_wrong_accumulation` and `_test_create_mocked_supervised_trainer` and the second is parametrizing `gradient_accumulation_steps` and calling others

Comment (vfdev-5) [reply tests/ignite/engine/test_create_supervised.py:372]: > Another solution could be split test_create_supervised_trainer into 2 tests such that the first test is executing once _test_create_supervised_trainer_wrong_accumulation and _test_create_mocked_supervised_trainer and the second is parametrizing gradient_accumulation_steps and calling others

@sadra-barikbin can you make suggested changes please ?

Comment (sadra-barikbin) [reply tests/ignite/engine/test_create_supervised.py:372]: Excuse me for delay. Sure.

Comment (sadra-barikbin) [reply tests/ignite/engine/test_create_supervised.py]: @vfdev-5 , I did the first one but also did some refactors. Does it look good?

Comment (vfdev-5) [reply tests/ignite/engine/test_create_supervised.py]: It's difficult to say as this adds more lines than it removes :)
Can you pull from your branch locally and push an empty commit to restart the CI. Autofix pep does not launch everything somehow...

Comment (sadra-barikbin) [reply tests/ignite/engine/test_create_supervised.py]: It's because of applying autopep(It has put each parameter of `pytest.param`s in a separate line). Now there is no redundancy.

Comment (sadra-barikbin) [reply tests/ignite/engine/test_create_supervised.py]: @vfdev-5 , what else should I do?

**Review by vfdev-5 — APPROVED** [review]
Thanks @sadra-barikbin , lgtm

**Comment by vfdev-5** [comment]
@sadra-barikbin your PR is failing on CUDA:
- https://app.circleci.com/pipelines/github/pytorch/ignite/2573/workflows/1e1e6a3c-95fb-45af-808a-2191b36788ab/jobs/7784
or
- https://github.com/pytorch/ignite/runs/6248177618?check_suite_focus=true#step:10:186
